perspective.el

```diff
@@ -781,22 +757,22 @@ See also `persp-add-buffer'."
   "Preserve the current perspective when entering a recursive edit."
   (persp-protect
     (persp-save)
-    (persp-frame-local-let ((persp-recursive persp-curr))
-      (let ((old-hash (copy-hash-table perspectives-hash)))
-        ad-do-it
-        ;; We want the buffer lists that were created in the recursive edit,
-        ;; but not the window configurations
-        (maphash (lambda (key new-persp)
-                   (let ((persp (gethash key old-hash)))
-                     (when persp
-                       (setf (persp-buffers persp) (persp-buffers new-persp)))))
-                 perspectives-hash)
-        (setq perspectives-hash old-hash)))))
+    (set-frame-parameter nil 'persp-recursive (frame-parameter nil 'persp-curr))
```

Comment:
Replacing `let` with `set` doesn't preserve the existing behavior--the frame parameter won't be reset after the function is called, or if it produces an exception.

I suggest replacing `persp-frame-local-let` with a new `persp-let-frame-parameter` macro that sets frame parameters for the duration of the body, and gracefully handles errors like `persp-frame-local-let`.

Reply:
Added a new macro `persp-let-frame-parameters`, inspired by `persp-frame-local-let`.